codec/src/test/java/io/netty/handler/codec/compression/Lz4FrameEncoderTest.java

```diff
@@ -52,4 +82,128 @@ protected ByteBuf decompress(ByteBuf compressed, int originalLength) throws Exce
+        final int blockSize = 100;
+        testAllocateBuffer(blockSize, blockSize - 13, false);
+        testAllocateBuffer(blockSize, blockSize * 5, false);
+        testAllocateBuffer(blockSize, NONALLOCATABLE_SIZE, false);
+    }
+
+    private void testAllocateBuffer(int blockSize, int bufSize, boolean isDirect) {
+        // allocate the input buffer to an arbitrary size less than the blockSize
+        ByteBuf in = ByteBufAllocator.DEFAULT.buffer(bufSize, bufSize);
+        in.writerIndex(in.capacity());
+
+        ByteBuf out = null;
+        try {
+            Lz4FrameEncoder encoder = newEncoder(blockSize, Lz4FrameEncoder.DEFAULT_MAX_ENCODE_SIZE);
+            out = encoder.allocateBuffer(ctx, in, isDirect);
+            Assert.assertNotNull(out);
+            if (NONALLOCATABLE_SIZE == bufSize) {
+                Assert.assertFalse(out.isWritable());
+            } else {
+                Assert.assertTrue(out.writableBytes() > 0);
+                Assert.assertEquals(isDirect, out.isDirect());
+            }
+        } finally {
+            in.release();
+            if (out != null) {
+                out.release();
+            }
+        }
+    }
+
+    @Test (expected = EncoderException.class)
+    public void testAllocateDirectBuffer_ExceedMaxEncodeSize() {
+        final int maxEncodeSize = 1024;
+        Lz4FrameEncoder encoder = newEncoder(Lz4Constants.DEFAULT_BLOCK_SIZE, maxEncodeSize);
+        int inputBufferSize = maxEncodeSize * 10;
+        ByteBuf buf = ByteBufAllocator.DEFAULT.buffer(inputBufferSize, inputBufferSize);
+        try {
+            buf.writerIndex(inputBufferSize);
+            encoder.allocateBuffer(ctx, buf, false);
+        } finally {
+            buf.release();
+        }
+    }
+
+    private Lz4FrameEncoder newEncoder(int blockSize, int maxEncodeSize) {
+        Checksum checksum = XXHashFactory.fastestInstance().newStreamingHash32(DEFAULT_SEED).asChecksum();
+        Lz4FrameEncoder encoder = new Lz4FrameEncoder(LZ4Factory.fastestInstance(), true,
+                                                      blockSize,
+                                                      checksum,
+                                                      maxEncodeSize);
+        encoder.handlerAdded(ctx);
+        return encoder;
+    }
+
+    @Test (expected = EncoderException.class)
+    public void testAllocateDirectBuffer_OverflowsOutputSize() {
+        final int maxEncodeSize = Integer.MAX_VALUE;
+        Lz4FrameEncoder encoder = newEncoder(Lz4Constants.DEFAULT_BLOCK_SIZE, maxEncodeSize);
+        ByteBuf buf = null;
+        try
+        {
+            buf = ByteBufAllocator.DEFAULT.buffer(maxEncodeSize, maxEncodeSize);
+            buf.writerIndex(maxEncodeSize);
+            encoder.allocateBuffer(ctx, buf, false);
+        } catch (OutOfDirectMemoryError e) {
+            // this sometimes fails on the continuous integration server if we try to allocate Integer.MAX_VALUE.
```

Comment:
This is to test the integer overflow aspects of `allocateBuffer()`, and the only way to get into that situation is to have a buffer near `Integer.MAX_VALUE`. Anecdotally, looks like allocating on-heap buffers did not have a problem and only direct buffers failed. Since it doesn't matter which buffer type gets us into the overflow situation, I will remove the large allocation for the direct buffer and rely on the on-heap allocation to test that path.